langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/stdin/StdioMcpTransport.java

```diff
@@ -0,0 +1,132 @@
+
+    private final List<String> command;
+    private final Map<String, String> environment;
+    private Process process;
+    private ProcessIOHandler processIOHandler;
+    private final Map<Long, CompletableFuture<JsonNode>> pendingOperations = new ConcurrentHashMap<>();
+    private final boolean logEvents;
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final Logger log = LoggerFactory.getLogger(StdioMcpTransport.class);
+
+    public StdioMcpTransport(Builder builder) {
+        this.command = builder.command;
+        this.environment = builder.environment;
+        this.logEvents = builder.logEvents;
+    }
+
+    @Override
+    public void start() {
+        log.info("Starting process: {}", command);
+        ProcessBuilder processBuilder = new ProcessBuilder(command);
+        processBuilder.environment().putAll(environment);
+        try {
+            process = processBuilder.start();
+            log.debug("PID of the started process: {}", process.pid());
+            process.onExit().thenRun(() -> {
+                log.debug("Subprocess has exited with code: {}", process.exitValue());
+            });
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        processIOHandler = new ProcessIOHandler(process, pendingOperations, logEvents);
+        // FIXME: where should we obtain the thread?
+        new Thread(processIOHandler).start();
+        new Thread(new ProcessStderrHandler(process)).start();
+    }
+
+    @Override
+    public JsonNode initialize(McpInitializeRequest request) {
+        try {
+            String requestString = OBJECT_MAPPER.writeValueAsString(request);
+            return executeAndWait(requestString, request.getId());
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public JsonNode listTools(McpListToolsRequest operation) {
+        try {
+            String requestString = OBJECT_MAPPER.writeValueAsString(operation);
+            return executeAndWait(requestString, operation.getId());
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public JsonNode executeTool(McpCallToolRequest operation) {
+        try {
+            String requestString = OBJECT_MAPPER.writeValueAsString(operation);
+            return executeAndWait(requestString, operation.getId());
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        process.destroy();
+    }
+
+    private JsonNode executeAndWait(String request, Long id) {
+        try {
+            CompletableFuture<JsonNode> future = new CompletableFuture<>();
+            pendingOperations.put(id, future);
+            processIOHandler.submit(request);
+            return future.get();
```

Comment:
no timeout

Reply:
That's a bit tricky because the user might not know how long a tool will take to finish... 
But right, maybe we could implement a timeout for tool executions, and send the server a cancellation message if the timeout is reached! I think that makes sense

Reply:
Java structured concurrency from JDK 21 preview would fit well here. Meanwhile, would it make sense to add timeout as a configuration parameter for the transport and set some meaningful default value, like 5 seconds?  Or keep it optional (nullable), but have an option for the user to set it.
nit: I would love to see JSpecify nullability annotations so the contract is more clear for the developer and the compiler

Reply:
Yeah unfortunately we're not on Java 21. Also, I think configurable tool timeouts are something that we should probably add in general, also for locally executed tools. So maybe it's something that we should keep as a separate task after all

Reply:
It's not that trivial because we should also send a cancellation message to the server that it can stop execution of this tool (the protocol supports this). Let's keep it as a follow-up

Reply:
What will happen if we don't send the cancellation message? Do servers support "rollback" functionality?

Reply:
I think nothing will really happen, except that the server will keep processing something that we don't need (or may be deadlocked), but I'd like to implement the cancellation properly